proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
+At run time, this signature is checked to match the signature of the function referenced by the provided `funcref`, trapping otherwise.
+This provides a dynamically typed form of function calls.
 
-`call_indirect` imposes a number of challenges (each discussed in more detail below), both currently and with respect to how WebAssembly might evolve:
-1. Looking at a function's definition alone, it's not apparent if a function can be indirectly called via `call_indirect`.
-One has to do some analysis to check if a reference to that function is made.
-And once a function reference has been made, it can be very difficult to ensure even something as simple as the function is only accessed by its own module.
-This makes it difficult to optimize functions or to ensure they are accessed in a way that respects the module's intended invariants.
-2. If subtyping is added, the signature provided by the indirect caller can be compatible with that of the callee function *without* being identical.
-Unfortunately, structural comparison of signatures at run time is likely too expensive.
-[WebAssembly/function-references#33](https://github.com/WebAssembly/function-references/pull/33) addresses this by requiring signatures to match exactly, but implementers of languages using subtyping have indicated that such a restriction does not suit their needs (e.g. subclasses in Kotlin can refine the signature of superclass methods).
-3. To practically support type imports/exports, `call_indirect` needs to compare caller and callee signatures using the "run-time value" of imported/exported types.
-This can be exploited to convert imported/exported types to and from the type they are supposed to be abstracting.
-Thus `call_indirect` can be used to access information that would otherwise be concealed behind an abstract type, or to create values of an abstract type that would otherwise be unforgeable.
+Call tags are a generalization and extension of this construct.
+The instruction `call_with_tag $call_tag` takes a `funcref` and calls it with the specified call tag.
+The type of the call tag indicates the param and result types of the function.
+`call_indirect $table $functype` is simply the special case `call_with_tag (call_tag.canon $functype) (table.get $table)`, calling the `funcref` fetched from the specified table with the *canonical* call tag for a given function signature.
 
-## Solution
+Every `func` definition (in theory) has a corresponding `funcref` that is returned by `ref.func`.
+By default, this `funcref` handles the canonical call tag for the functions signature.
+But with this extension, a `func` definition can alternatively specify which call tags the `funcref` should handle (so long as their types are compatible with the function's signature).
+These can be *multiple* canonical call tags, and they can include custom call tags either generated by (and specific to) this module instance or imported from other module instances.
```

Comment:
By "multiple canonical call tags," do you mean canonical call tags for multiple signatures? I assume those signatures would have to be subtypes of the `func`'s signature.